### PR TITLE
Factorize code between Searchlight and NiftiSpheresMasker

### PR DESCRIPTION
**Pull request by AlexandreAbraham**
Addresses problem 2 in #720.

### CODE REVIEW

File: nilearn/input_data/nifti_spheres_masker.py [hunk 1]
```diff
@@ -17,33 +17,8 @@
...
-
-    allow_overlap: boolean
-        If False, an error is raised if the maps overlaps (ie at least two
-        maps have a non-zero value for the same voxel). Default is False.
-
-    mask_img: Niimg-like object, optional
-        See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
-        Mask to apply to regions before extracting signals.
-    """
+def _apply_mask_and_get_affinity(seeds, niimg, radius, allow_overlap,
```
Comment (mrahim) [inline nilearn/input_data/nifti_spheres_masker.py]: I don't know if `affinity` might not be misleading (e. g. affine) and if `adjacency` would be clearer.

File: nilearn/input_data/nifti_spheres_masker.py [hunk 2]
```diff
@@ -82,12 +57,39 @@ def _iter_signals_from_spheres(seeds, niimg, radius, allow_overlap,
...
+    
+    return X, A
 
+
+def _iter_signals_from_spheres(seeds, niimg, radius, allow_overlap,
+                               mask_img=None):
+    """Utility function to iterate over spheres.
+    Parameters
+    ==========
+    seeds: List of triplet of coordinates in native space
```
Comment (mrahim) [inline nilearn/input_data/nifti_spheres_masker.py]: ... List of triplet**s** ...

File: nilearn/input_data/nifti_spheres_masker.py [hunk 3]
```diff
@@ -82,12 +57,39 @@ def _iter_signals_from_spheres(seeds, niimg, radius, allow_overlap,
...
+        Seed definitions. List of coordinates of the seeds in the same space
+        as the images (typically MNI or TAL).
+    imgs: 3D/4D Niimg-like object
+        See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
```
Comment (mrahim) [inline nilearn/input_data/nifti_spheres_masker.py]: replace URL by http://nilearn.github.io/manipulating_visualizing/manipulating_images.html#niimg

Comment (lesteve) [inline nilearn/input_data/nifti_spheres_masker.py]: Interesting, that means that all the current links in our docstring are broken.

File: nilearn/input_data/nifti_spheres_masker.py (hunk 1 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_spheres_masker.py]: Affinity is more general. However, this is a private function, I think it's not a big deal.

File: nilearn/input_data/nifti_spheres_masker.py (hunk 2 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_spheres_masker.py]: Fixed.

File: nilearn/input_data/nifti_spheres_masker.py (hunk 3 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_spheres_masker.py]: This should be done in a dedicated PR.

**Comment by GaelVaroquaux** [comment]
@AlexandreAbraham : should we try to merge this today?

**Comment by AlexandreAbraham** [comment]
For me, it's ready to merge. It's really just a factorization of 2 identical pieces of code.

File: nilearn/decoding/searchlight.py [hunk 4]
```diff
@@ -279,36 +279,24 @@ def fit(self, imgs, y):
...
+
+        # Compute world coordinates of the seeds
+        process_mask, process_mask_affine = masking._load_mask_img(
+            process_mask_img)
+        process_mask_coords = np.where(process_mask != 0)
+        process_mask_coords = \
+            np.asarray(process_mask_coords +
+                       (np.ones(len(process_mask_coords[0]), dtype=np.int),))
+        process_mask_coords = np.dot(process_mask_affine,
+                                     process_mask_coords)[:3].T
```
Comment (bthirion) [inline nilearn/decoding/searchlight.py]: This manipulation is probably done in several places. Maybe it could be factorized too.

**Comment by bthirion** [comment]
LGTM. I am only suggesting some further factorization.

File: nilearn/decoding/searchlight.py (hunk 4 above)
Comment (GaelVaroquaux) [inline nilearn/decoding/searchlight.py]: > > +        process_mask_coords = np.dot(process_mask_affine,
> > +                                     process_mask_coords)[:3].T

> This manipulation is probably done in several places. Maybe it could be
> factorized too.

Can nilearn.image.resampling.coord_transform be used here?

Comment (AlexandreAbraham) [inline nilearn/decoding/searchlight.py]: Yes. I think there are a lot of places in nilearn where this change hasn't been done. I could do a PR to propagate it.

Comment (GaelVaroquaux) [inline nilearn/decoding/searchlight.py]: > Yes. I think there are a lot of places in nilearn where this change
> hasn't been done. I could do a PR to propagate it.

How about you do it here, we merge this PR, and you do another PR to
propagate it?

Comment (AlexandreAbraham) [inline nilearn/decoding/searchlight.py]: Fixed in the last commit. Also replaced an instance of this code in `nifti_spheres_masker`.

File: nilearn/decoding/searchlight.py [hunk 5]
```diff
@@ -279,36 +280,23 @@ def fit(self, imgs, y):
...
+        process_mask, process_mask_affine = masking._load_mask_img(
+            process_mask_img)
+        process_mask_coords = np.where(process_mask != 0)
+        process_mask_coords = coord_transform(
+            process_mask_coords[0], process_mask_coords[1],
+            process_mask_coords[2], process_mask_affine)
+        process_mask_coords = np.asarray(process_mask_coords).T
+
+        X, A = _apply_mask_and_get_affinity(
+            process_mask_coords, imgs, self.radius, True,
```
Comment (GaelVaroquaux) [inline nilearn/decoding/searchlight.py]: Could you use a bit more keyword arguments: I have no idea what this 'True' hanging around is doing.

**Comment by GaelVaroquaux** [comment]
LGTM. Great job.

I am +1 on this. @bthirion is too. I am merging!
